### PR TITLE
NTT-10 | Added connector, service, controller

### PR DESCRIPTION
**Pull request by owen-opencast**

### CODE REVIEW

File: app/models/Validation.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  private val surnamePattern = "^(?=.{1,99}$)([A-Z]([-'. ]{0,1}[A-Za-z ]+)*[A-Za-z]?)$".r
+
+  private val forenamePattern = "^(?=.{1,99}$)([A-Z]([-'. ]{0,1}[A-Za-z ]+)*[A-Za-z]?)$".r
+
+  private val birthdatePattern = """^(((19|20)([2468][048]|[13579][26]|0[48])|2000)[-]02[-]29|((19|20)[0-9]{2}[-](0[469]
+                                   ||11)[-](0[1-9]|1[0-9]|2[0-9]|30)|(19|20)[0-9]{2}[-](0[13578]|1[02])[-](0[1-9]|[12][0-9]|3[01])|(19|20)[0-9]{2}
+                                   |[-]02[-](0[1-9]|1[0-9]|2[0-8])))$""".r
+
+  val NinoReads: Reads[String] = Reads.pattern(ninoPattern, "Not a valid National Insurance number")
```
Comment (ajchilts) [inline app/models/Validation.scala:33]: We need to be careful these errors do not get logged or fed back to the user? We should always reply with the something went wrong error.

File: app/services/IdentityMatchService.scala [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+class IdentityMatchService @Inject()(val connector: IdentityMatchConnector,
+                                     val repository: IndividualCheckRepository,
+                                     val appConfig: AppConfig) {
+
+
+  def matchId(request: IdMatchRequest)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Either[IdMatchError, IdMatchResponse]] = {
+
+    limitedMatch(request).recoverWith {
+      case ex: LimitException => getErrorResponse(ex.getLocalizedMessage)
+      case _: InvalidIdMatchResponse => getErrorResponse("Something went wrong")
```
Comment (ajchilts) [inline app/services/IdentityMatchService.scala:39]: Ah - you have replied with the generic error 👍

Comment (owen-opencast) [reply app/services/IdentityMatchService.scala:39]: Yeah, I thought if I handled them explicitly it would be easier to expand the code in future 🙂

**Comment by adamconder** [comment]
Hi all, I had a couple of questions:

What do you see making up the id property as part of? `final case class IdMatchRequest(id: String, nino: String, surname: String, forename: String, birthDate: String)`. My suspicions would be internalId from auth? If this is the case does mongo have an index to ensure it is a _compound index_ between the trust identifier (UTR/URN) and auth id? We need to ensure this for Agents to be able to register multiple trusts at one time, without locking out their credential.

I see you have implemented a `clearLock` function, how did you envisage this being used in the journeys?

**Comment by ajchilts** [comment]
Hi @adamconder 

So for the 'Id' we had not looked that far into the journey. We have built this as a stand alone piece for the time being. The Id could be whatever you deem fit at the point of piecing together that journey. I initially thought about the Internal Id from auth however; how does this gel if the user has multiple trusts each with a different lead trustee? You would not be able to lock out on say trust two based on an individual check fail from trust one right? So this Id (hence it simply being called 'Id' at this stage) should probably be specific to the trust and not the credentials themselves? If that makes sense? 

I would envisage the clear method being used if a user manages to pass the individual check on a subsequent attempt. I.e. Let us say a user gets two failed attempts and succeed on the third. Would we not want to clear the attempts at this stage? Ball totally in your court on that one however; we decided to add this functionality.

**Comment by adamconder** [comment]
@ajchilts Thanks for the reply!

I agree with your comments on the id, without changing the identifier implementation, like you say we can chose to use it how we wish. 👍

**Pull request merged**